apps/designer/app/canvas/shared/use-drag-drop.ts

```diff
@@ -0,0 +1,273 @@
+    }
+
+    return { element, data: rootInstance };
+  };
+
+  const dropHandlers = useDrop<Instance>({
+    isDropTarget(element) {
+      return (
+        (rootInstance !== undefined &&
+          element.id !== "" &&
+          findInstanceById(rootInstance, element.id)) ||
+        false
+      );
+    },
+
+    // This must be fast, it can be called multiple times per pointer move
+    swapDropTarget(dropTarget) {
+      const { dragItem } = state.current;
+
+      if (!dropTarget || dragItem === undefined || rootInstance === undefined) {
+        return getDefaultDropTarget();
+      }
+
+      if (dropTarget.data.id === rootInstance.id) {
+        return dropTarget;
+      }
+
+      const path = getInstancePath(rootInstance, dropTarget.data.id);
+      path.reverse();
+
+      if (dropTarget.nearEdge) {
+        path.shift();
+      }
+
+      // Don't allow to dpop inside drag item or any of its children
+      const dragItemIndex = path.findIndex((x) => x.id === dragItem.id);
+      if (dragItemIndex !== -1) {
+        path.splice(0, dragItemIndex + 1);
+      }
+
+      const data = path.find((instance) =>
+        primitives[instance.component].canAcceptChild()
+      );
+
+      if (!data) {
+        return getDefaultDropTarget();
+      }
+
+      if (data.id === dropTarget.data.id) {
+        return dropTarget;
+      }
+
+      const element = data && document.getElementById(data.id);
+
+      if (!element) {
+        return getDefaultDropTarget();
+      }
+
+      return { data, element };
+    },
+
+    onDropTargetChange(dropTarget) {
+      state.current.dropTarget = dropTarget;
+      publish<"dropTargetChange", DropTargetChangePayload>({
+        type: "dropTargetChange",
+        payload: {
+          rect: dropTarget.rect,
+          placement: dropTarget.placement,
+          position: dropTarget.indexWithinChildren,
+          instanceId: dropTarget.data.id,
+          instanceComponent: dropTarget.data.component,
+        },
+      });
+    },
+  });
+
+  const dragHandlers = useDrag<Instance>({
+    isDragItem(element) {
+      if (rootInstance === undefined || element.id === "") {
+        return false;
+      }
+
+      const instance = findInstanceById(rootInstance, element.id);
+
+      if (instance === undefined) {
+        return false;
+      }
+
+      // We can't drag if we are editing text
+      if (instance.id === textEditingInstanceId) {
+        return false;
+      }
+
+      // Cannot drag root
+      if (instance.id === rootInstance.id) {
+        return false;
+      }
+
+      return instance;
+    },
+    onStart({ data: instance }) {
+      state.current.dragItem = instance;
+
+      autoScrollHandlers.setEnabled(true);
+
+      publish<"dragStart", DragStartPayload>({
+        type: "dragStart",
+        payload: {
+          origin: "canvas",
+          dragItem: instance,
+        },
+      });
+    },
+    onMove: (poiterCoordinate) => {
```

Comment:
there must be a better name than pointerCoordinate, also there is a typo, maybe just `point`?